File: 6. ZigZag Conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows == 1:
            return s
        ret = [[] for i in range(numRows)]
        n = len(s)
        i = 0
        while i < n:
            for j in range(numRows):
                if i + j < n:
                    ret[j].append(s[i + j])
            i += numRows
            for j in range(numRows-2):
                if i + j < n:
                    ret[numRows -2 -j].append(s[i + j])
            i += numRows - 2
        return ''.join(map(lambda x: ''.join(x), ret))

File: 6. Zigzag Conversion.go
func convert(s string, numRows int) string {
	n := len(s)
	if n <= numRows || numRows == 1 {
		return s
	}
	m := numRows*2 - 2
	res := make([]byte, 0, n)
	for i := 0; i < numRows; i++ {
		for j := i; j < n; j += m {
			res = append(res, s[j])
			if i != 0 && i != numRows-1 && j+m-2*i < n {
				res = append(res, s[j+m-2*i])
			}
		}
	}
	return string(res)
}